usaspending_api/etl/tests/test_usaspending_assistance_load.py

```diff
@@ -19,6 +22,72 @@ def test_usaspending_assistnace_load():
     # @todo - should there be an assert here?
 
 
+@pytest.mark.django_db(transaction=True)
+def test_award_and_txn_uniqueness():
+    """Each Award should be unique by:
+    - PIID (Contract) or FAIN/URI (Grants)
+    - Parent Award (but grants don't have parents)
+    - Awarding sub-tier Agency
+
+    Each Transaction should be unique by
+    - Award
+    - Modification number"""
+    call_command('loaddata', 'endpoint_fixture_db')
+
+    # agency records from reference_fixture are a prereq
+
+    awards_before_loads = Award.objects.count()
+    txn_before_loads = Transaction.objects.count()
+    call_command('loaddata', 'reference_fixture')
+    filepath = os.path.join(settings.BASE_DIR, 'usaspending_api', 'data',
+                            'usaspending_fin_assist_direct_payments.csv')
+    call_command('load_usaspending_assistance', filepath)
+    base_award_count = Award.objects.count() - awards_before_loads
+    base_txn_count = Transaction.objects.count() - txn_before_loads
+
+    # Re-importing with new FAINS makes new awards
+    def fainchange(row):
+        return _replacer(row, 'federal_award_id', fain_prefixer)
+
+    def new_fain(row):
+        row['federal_award_id'] = '{}-fain2'.format(row['federal_award_id'])
+        return row
+
+    with mutated_csv(filepath, new_fain) as mutant_file:
+        call_command('load_usaspending_assistance', mutant_file.name)
+        print('ok')
+    assert Award.objects.count() == awards_before_loads + (2 * base_award_count
+                                                           )
+
+    # Find an agency with no awards
+    agencies_in_use = Award.objects.values('awarding_agency__id')
+    new_agency = Agency.objects.exclude(id__in=agencies_in_use).first()
+    # Assign loaded records to that agency
+    for award in Award.objects.all():
+        award.awarding_agency = new_agency
+        award.save()
+    # re-import original file - should have new records
+    call_command('load_usaspending_assistance', filepath)
+    assert Award.objects.count() == awards_before_loads + (3 * base_award_count
```

Comment:
Again, take this with a grain of salt coming from a reviewer who skipped breakfast, but I had to pause and think why the post-awarding agency update award count would be 3 times the original number instead of 2 times (of course, it's because of the previous test). From a code reader perspective, it might be helpful to have more narrative in the comments (_e.g._, "changing FAINS on the input file and re-running the load should create a new award for each record"...."changing awarding agency in the db and re-running the load should cause an additional set of new awards to be created).

Again, not needed for merging, and the most important thing is having these robust tests around the tricky logic for creating awards. I really like the way you've done this!